Implement mapped value unpacking

uranusjr opened the pull request:
This is another PR that is likely much too big. Many (sort of interrelated) things done in one.

#### Rewrite `@task` mapped operators expansion so it behave correctly

Previously, a mapped `@task` was not expanded correctly because its mapped argument would look something like this:

```python
mapped_arguments={
    "op_kwargs": {
        "arg": [1, 2, 3],
    },
}
```

this cannot be expanded using normal logic, because `PythonOperator` natively expects something like this instead:

```python
mapped_arguments={
    "op_kwargs": [
        {"arg": 1},
        {"arg": 2},
        {"arg": 3},
    ],
}
```

so additional logic is implemented to modify `expand_mapped_task` to correctly expand.

#### Re-implement `expand_mapped_task` logic to be able to expand literals

Previously, `expand_mapped_task` only looks in TaskMap and cannot expand e.g. `.map(arg=[1, 2, 3])`. This implements a more sophisticated logic to collect information from mapped arguments and expand literal and XComArg inputs properly. Some refactoring was also done so the same logic can be reused for task-runtime value unpacking.

#### Additional checks added to ensure correct literal types are passed to `.map()`

Basically making sure it only receives lists and dicts (or XComArg, which is already checked separately when the upstream pushes to XCom).

#### Extend `TaskInstance.render_templates` to “unpack” mapped values for task execution

This is the main thing. After all template fields are rendered, additional process is done to “unpack” mapped values into individual ones based on map_index, which further mutates the operator object held by a TaskInstance. This reuses similar logic from `expand_mapped_task` to calculate the total length of map, so it can locate its map_index inside the series.

Previously, `render_template_fields` was only implemented on BaseOperator, and a MappedOperator is first unmapped before templates are rendered. This approach was unforuantely wrong, since value unpacking (which needs to happen after template rendering resolves XComArg) needs to be aware of the original MappedOperator (mainly to access the user-supplied mapped kwargs). So the new logic delays unmapping until _during_ `render_templates`. If a TaskInstance’s `task` is a MappedOperator, its `render_templates` would call the MappedOperator’s `render_template_fields`, which unmaps itself, calls `render_template_fields` on the unmapped operator, unpacks values for the unmapped operator, _and then returns the unmapped operator_. The TaskInstance would reassign its `task` to the unmapped, rendered, unpacked operator. This means that `TaskInstance.render_templates` now may have a side effect of mutating its own `task` attribute. I don’t particularly like this, but couldn’t find a cleaner approach without breaking much of the existing interface (complicated by the fact that both `BaseOperator.render_template_fields` and `BaseOperator.render_template` are public API). I think this is close to the best possible interface considering existing constraints (and I try to document this as clearly as possible for future maintainability).

A few tests were added to check this is working (see `TestMappedTaskInstanceReceiveValue`).

uranusjr committed:
Do not map op_args to empty list

uranusjr committed:
Fixup how we expand task

This new expansion logic should cover literal expansions and expanding
against multiple map() arguments. I think the logic is still not quite
right (e.g. mapping against a *mapped* upstream, and *when* we map
exactly), but this new logic should be useful for what we're going to
need to fix anyway.

uranusjr committed:
Implement mapped value unpacking

uranusjr committed:
Fix runtime import in TaskInstance

uranusjr committed:
Improve SnowflakeToSlack tests to properly clean up

uranusjr committed:
Fix BaseOperator.render_template_fields compat

This method is public API and may be subclassed, so we can't change its
semantics, specifically we need to handle it returning None (as is the
case currently in main).

uranusjr committed:
Remove stray class left over from refactoring

uranusjr committed:
Rename argument/variable "content" to "value"

The previous name looks too much like "context" and is an unnecessary
hinderance to maintenance.

uranusjr committed:
Make test_mapped_classic actually work

uranusjr committed:
Document our validate-by-mocking approach

uranusjr committed:
Some more runtime validation on mapped length

uranusjr committed:
Add test for expanding product-map

ashb commented on airflow/models/mappedoperator.py:
```diff
@@ -55,38 +58,65 @@
+    # mapping since we need the value to be ordered).
+    MapArgument = Union[XComArg, Sequence, dict]
+
+ValidationSource = Union[Literal["map"], Literal["partial"]]
```
I think you can do this

```suggestion
ValidationSource = Literal["map", "partial"]
```

uranusjr replied:
Unfortunately this only works on 3.9+.

ashb replied:
Ahhh

ashb commented on airflow/decorators/base.py:
```diff
@@ -269,22 +276,21 @@ def __call__(self, *args, **kwargs) -> XComArg:
+            value = kwargs_left.pop(arg_name, NOTSET)
+            if func != "map" or value is NOTSET or isinstance(value, get_mappable_types()):
+                continue
+            raise ValueError(f"{func} got unexpected value{type(value)!r} for keyword argument {arg_name!r}")
```
Shouldn't this be a TypeError? (That's why Python throws in this situation)

ashb replied:
🤦🏻 I see. Ignore me.

ashb commented on airflow/models/abstractoperator.py:
```diff
@@ -17,14 +17,18 @@
-from typing import TYPE_CHECKING, Any, Callable, Collection, Dict, List, Optional, Set, Type, Union
+from typing import TYPE_CHECKING, Any, Callable, Collection, Dict, Iterable, List, Optional, Set, Type, Union
+
+from sqlalchemy.orm import Session
```
Type checking only import?

uranusjr replied:
Yes, but if `Session` is ever passed in as an argument this class is already in memory anyway, making this import basically free.

ashb commented on airflow/models/abstractoperator.py:
```diff
@@ -249,3 +253,124 @@ def get_extra_links(self, dttm: datetime.datetime, link_name: str) -> Optional[D
+            rendered_content = self._render_template_field(
+                attr_name,
+                content,
+                context,
```
nit: `context` and `content` are very similar, and we'd be somewhat error prone to confuse them. Can we rename `content` to something like `template`?

uranusjr replied:
Good idea, will do. (This was cut-pasted from BaseOperator.)

uranusjr replied:
We unfortunately can’t rename the argument name in `render_template()` since it’s public API. I renamed `content` to `value` everywhere else. I avoided `template` since it is used to refer to the compiled Jinja template object.

ashb commented on airflow/models/mappedoperator.py:
```diff
@@ -475,37 +517,119 @@ def expand_mapped_task(
+
+
+class _FieldNotMapped(Exception):
+    """Raised by _expand_mapped_field if a field is not mapped."""
```
Not used currently?

uranusjr replied:
Forgot to delete it after refactoring.

ashb commented on airflow/decorators/base.py:
```diff
@@ -369,31 +371,43 @@ class DecoratedMappedOperator(MappedOperator):
+        if real:
+            mapped_op_kwargs: Dict[str, Any] = self.mapped_op_kwargs
+        else:
+            mapped_op_kwargs = {k: unittest.mock.MagicMock(name=k) for k in self.mapped_op_kwargs}
```
This probably needs a huge comment saying what we're doing 😁

uranusjr replied:
I extracted this into a function with a three-paragraph docstring.

uranusjr committed:
Removing from DAG removes from task group

So we don't need to remove the task from the group ourselves.

uranusjr committed:
Sort tis by map_index to test stricter value order

uranusjr committed:
Fix TaskMap session management in test

uranusjr committed:
Allow a task to provide multiple mapping arguments

Since a task can potentially be referred to multiple times by a
downstream, the task-arg relation needs to be a multimap instead of a
simple one-to-one.

ashb commented on airflow/models/mappedoperator.py:
```diff
@@ -245,16 +296,13 @@ def _validate_argument_count(self) -> None:
-            operator.task_group._remove(operator)
-        dag = operator.get_dag()
+        mocked_mapped_kwargs = create_mocked_kwargs(self.mapped_kwargs)
+        op = self._create_unmapped_operator(mapped_kwargs=mocked_mapped_kwargs, real=False)
```
I do wonder if we'll need an escape hatch for this.

I guess we can add that when someone finds a specific case.

uranusjr replied:
Escape hatch for if an operator cannot correctly run `__init__` with mocks?

I think eventually we may want to create a separate API for validation without actually creating the operator, perhaps a classmethod called `validate` that can be called here.

ashb replied:
Yeah, I was thinking an escape hatch for when an Operator ctor does something that causes it to fail when given a Mock.

A proper class-level `validate(**kwargs)` sounds like a better idea.

ashb commented on airflow/models/mappedoperator.py:
```diff
@@ -103,6 +134,24 @@ def prevent_duplicates(kwargs1: Dict[str, Any], kwargs2: Dict[str, Any], *, fail
+    since it not only provide good run-time properties, but also enjoy special
+    treatments in Mypy.
+    """
+    return {k: unittest.mock.MagicMock(name=k) for k in kwargs}
```
Future possible enhancement (absolutely not in this pr):

Set `spec` for the mock based on the type hint of the param we are are creating the mock for.

ashb commented on airflow/models/mappedoperator.py:
```diff
@@ -245,16 +296,13 @@ def _validate_argument_count(self) -> None:
+        dag = op.get_dag()
         if dag:
-            dag._remove_task(operator.task_id)
+            dag._remove_task(op.task_id)
```
```python
        dag = op.get_dag()
        if dag:
            dag._remove_task(op.task_id)
```

I think that is all we need -- dag._remove_tasks does the taskgroup check/remove already.

ashb commented on tests/models/test_taskinstance.py:
```diff
@@ -2353,3 +2353,110 @@ def pull_something(value):
+            (2, ("b", "y")),
+            (1, ("c", "z")),
+            (2, ("c", "z")),
+        }
```
I don't think it matters, but I would have expected `emit_numbers to be the outer loop since it's the "first" argument

```python
        assert outputs == {
            (1, ("a", "x")),
            (1, ("c", "z")),
            (1, ("b", "y")),
            (2, ("a", "x")),
            (2, ("b", "y")),
            (2, ("c", "z")),
        }
```


i.e. `show.map(number=emit_numbers(), letter=emit_letters())` is the equiv of:
```python

for number in emit_numbers():
    for letter in emit_letters():
        show(number, letter)
```

uranusjr replied:
This is a set so the ordering doesn’t matter :p The actual ordering is correct if the task instances are ordered by `map_index`. (But that ordering is not guaranteed on the list of _task instances_ returned by `expand_mapped_task`; I can add additional logic to test that.)

ashb replied:
I think we need the order to be stable/guaranteed. What it is doesn't matter so much, just that it is stable and consistent. (We said we'd have this in the AIP)

uranusjr replied:
We guarantee that map_index would be generated in a predictable order, but not the mapped task instances would execute in that same order (which is the potentially unstable part here).

uranusjr replied:
With that said, I added a `sorted()` call and tweaked the tests to ensure stricter ordering.

ashb replied:
Yeah, execution order is not important just that map_index=0 is always the same values in the case of mapping over multiple TaskMaps.

ashb reviewed (approved):
LGTM. couple of possible changes, but those can be done later (or not at all)
